Source/Core/AudioCommon/SoundStream.cpp

```diff
@@ -0,0 +1,195 @@
+			sample = MathUtil::Clamp(sample, -32768.f, 32767.f);
+			dst[i * numchannels + j] = s16(sample);
+		}
+	}
+}
+
+__forceinline void s16ToFloat(float* dst, const s16 *src, u32 numsamples)
+{
+	for (u32 i = 0; i < numsamples; i++)
+	{
+		s16 sample = src[i];
+		dst[i] = Signed16ToFloat(sample);
+	}
+}
+
+// The audio thread.
+void SoundStream::SoundLoop()
+{
+	Common::SetCurrentThreadName("Audio thread");
+	InitializeSoundLoop();
+	bool surroundSupported = SupportSurroundOutput() && SConfig::GetInstance().bDPL2Decoder;
+	memset(realtimeBuffer, 0, SOUND_MAX_FRAME_SIZE * sizeof(u16));
+	memset(dpl2buffer, 0, SOUND_MAX_FRAME_SIZE * sizeof(soundtouch::SAMPLETYPE));
+	memset(samplebuffer, 0, SOUND_MAX_FRAME_SIZE * sizeof(soundtouch::SAMPLETYPE));
+	memset(s_dither_prev, 0, sizeof(s_dither_prev));
+	u32 channelmultiplier = surroundSupported ? SOUND_SAMPLES_SURROUND : SOUND_SAMPLES_STEREO;
+	CMixer* mixer = GetMixer();
+	if (SConfig::GetInstance().bTimeStretching)
+	{
+		float ratemultiplier = 1.0f;
+		soundtouch::SoundTouch sTouch;
+		sTouch.setChannels(2);
+		sTouch.setSampleRate(mixer->GetSampleRate());
+		sTouch.setTempo(1.0);
+		sTouch.setSetting(SETTING_USE_QUICKSEEK, 0);
+		sTouch.setSetting(SETTING_USE_AA_FILTER, 1);
+		while (threadData.load())
+		{
+			u32 availablesamples = mixer->AvailableSamples();
+			u32 numsamples = std::min(availablesamples, 400u);
+			if (numsamples == 400u)
+			{
+				float rate = mixer->GetCurrentSpeed();
+				if (rate <= 0)
+				{
+					rate = 1.0f;
+				}
+				numsamples = mixer->Mix(samplebuffer, numsamples);
+				rate *= ratemultiplier;
+				rate = rate < 0.5f ? 0.5f : rate;
+				rate = roundf(rate * 16.0f) / 16.0f;
+				sTouch.setTempo(rate);
+				sTouch.putSamples(samplebuffer, numsamples);
+			}
+			u32 samplesneeded = SamplesNeeded();
+			availablesamples = sTouch.numSamples();
+			if (samplesneeded >= SOUND_FRAME_SIZE && availablesamples > 0)
+			{
+				ratemultiplier = std::fmaxf(std::fminf((float)availablesamples / (float)samplesneeded, 1.1f), 0.9f);
+				numsamples = std::min(availablesamples, SOUND_FRAME_SIZE);
+				if (surroundSupported)
+				{
+					numsamples = sTouch.receiveSamples(dpl2buffer, numsamples);
+					DPL2Decode(dpl2buffer, numsamples, samplebuffer);
+					// zero-out the subwoofer channel - DPL2Decode generates a pretty
+					// good 5.0 but not a good 5.1 output.  Sadly there is not a 5.0
+					// AL_FORMAT_50CHN32 to make this super-explicit.
+					// DPL2Decode output: LEFTFRONT, RIGHTFRONT, CENTREFRONT, (sub), LEFTREAR, RIGHTREAR
+					for (u32 i = 0; i < numsamples; ++i)
+					{
+						samplebuffer[i*SOUND_SAMPLES_SURROUND + 3 /*sub/lfe*/] = 0.0f;
+					}
+				}
+				else
+				{
+					numsamples = sTouch.receiveSamples(samplebuffer, numsamples);
+				}
+				floatTos16(realtimeBuffer, samplebuffer, numsamples, channelmultiplier);
+				WriteSamples(realtimeBuffer, numsamples);
+			}
+			else
+			{
+				Common::SleepCurrentThread(1);
+			}
+		}
+	}
+	else
```

Comment:
This cases should be merged. There is no need to have two calls to DPL encoder, or mixer. Just time stretching should be optional.